File: scripts/manage-project-runtime.py
#!/usr/bin/env python3
import argparse,hashlib,json,os,shutil
from pathlib import Path
def sha(p):return hashlib.sha256(p.read_bytes()).hexdigest()
def replace_link(link,target):
 tmp=link.with_name('.'+link.name+'.tmp');tmp.unlink(missing_ok=True);tmp.symlink_to(target);os.replace(tmp,link)
def verify_release(root):
 m=json.loads((root/'.harness-runtime.json').read_text())
 for f in m['files']:
  p=root/f['path']
  if not p.is_file() or sha(p)!=f['sha256']:raise RuntimeError('runtime drift: '+f['path'])
 return m['digest']
def activate(runtime,project,digest):
 target=runtime/'releases'/digest
 if not target.is_dir():raise RuntimeError('runtime release missing: '+digest)
 verify_release(target);current=runtime/'current';old=current.resolve().name if current.is_symlink() else None
 if old:(runtime/'previous').write_text(old+'\n')
 replace_link(current,Path('releases')/digest);replace_link(project/'.agents',Path('build/harness/runtime/current/.agents'));replace_link(project/'.goose',Path('build/harness/runtime/current/.goose'))
def main():
 ap=argparse.ArgumentParser();ap.add_argument('action',choices=['activate','verify','rollback','clean']);ap.add_argument('--runtime-root',required=True);ap.add_argument('--project-root',required=True);a=ap.parse_args();runtime=Path(a.runtime_root).resolve();project=Path(a.project_root).resolve()
 if a.action=='activate':activate(runtime,project,(runtime/'candidate').read_text().strip())
 elif a.action=='verify':
  if not (runtime/'current').is_symlink():raise RuntimeError('runtime current missing')
  digest=verify_release((runtime/'current').resolve())
  for name in ('.agents','.goose'):
   link=project/name
   if not link.is_symlink() or link.resolve()!=(runtime/'current'/name).resolve():raise RuntimeError('project runtime link mismatch: '+name)
  print(digest)
 elif a.action=='rollback':
  previous=runtime/'previous'
  if not previous.exists():raise RuntimeError('no previous runtime')
  activate(runtime,project,previous.read_text().strip())
 else:
  current=(runtime/'current').resolve() if (runtime/'current').is_symlink() else None
  previous=(runtime/'previous').read_text().strip() if (runtime/'previous').exists() else None
  for p in (runtime/'releases').iterdir() if (runtime/'releases').exists() else []:
   if p!=current and p.name!=previous:shutil.rmtree(p)
if __name__=='__main__':main()

File: src/app/harness-manager/src/manage-project-runtime.ts
#!/usr/bin/env node
import { createHash } from "node:crypto";
import { existsSync, lstatSync, mkdirSync, readFileSync, readlinkSync, realpathSync, readdirSync, renameSync, rmSync, symlinkSync, writeFileSync } from "node:fs";
import { basename, dirname, join, resolve } from "node:path";
import type { ProjectionManifest } from "./project-harness-runtime.js";

function sha256File(path: string): string { return createHash("sha256").update(readFileSync(path)).digest("hex"); }
function isSymlink(path: string): boolean { try { return lstatSync(path).isSymbolicLink(); } catch { return false; } }

export function replaceLink(link: string, target: string): void {
  const temporary = join(dirname(link), "." + basename(link) + ".tmp");
  rmSync(temporary, { force: true });
  symlinkSync(target, temporary);
  renameSync(temporary, link);
}

export function verifyRelease(root: string): string {
  const manifest = JSON.parse(readFileSync(join(root, ".harness-runtime.json"), "utf8")) as ProjectionManifest;
  for (const file of manifest.files) {
    const path = join(root, file.path);
    if (!existsSync(path) || !lstatSync(path).isFile() || sha256File(path) !== file.sha256) throw new Error("runtime drift: " + file.path);
  }
  return manifest.digest;
}

export function activateRuntime(runtimeInput: string, projectInput: string, digest: string): void {
  const runtime = resolve(runtimeInput);
  const project = resolve(projectInput);
  const target = join(runtime, "releases", digest);
  if (!existsSync(target) || !lstatSync(target).isDirectory()) throw new Error("runtime release missing: " + digest);
  verifyRelease(target);
  const current = join(runtime, "current");
  const old = isSymlink(current) ? basename(realpathSync(current)) : undefined;
  if (old !== undefined) writeFileSync(join(runtime, "previous"), old + "\n");
  replaceLink(current, join("releases", digest));
  replaceLink(join(project, ".agents"), "build/harness/runtime/current/.agents");
  replaceLink(join(project, ".goose"), "build/harness/runtime/current/.goose");
}

export type RuntimeAction = "activate" | "verify" | "rollback" | "clean";
export function manageProjectRuntime(action: RuntimeAction, runtimeInput: string, projectInput: string): string | undefined {
  const runtime = resolve(runtimeInput);
  const project = resolve(projectInput);
  if (action === "activate") {
    activateRuntime(runtime, project, readFileSync(join(runtime, "candidate"), "utf8").trim());
  } else if (action === "verify") {
    const current = join(runtime, "current");
    if (!isSymlink(current)) throw new Error("runtime current missing");
    const digest = verifyRelease(realpathSync(current));
    for (const name of [".agents", ".goose"]) {
      const link = join(project, name);
      if (!isSymlink(link) || realpathSync(link) !== realpathSync(join(current, name))) throw new Error("project runtime link mismatch: " + name);
    }
    return digest;
  } else if (action === "rollback") {
    const previous = join(runtime, "previous");
    if (!existsSync(previous)) throw new Error("no previous runtime");
    activateRuntime(runtime, project, readFileSync(previous, "utf8").trim());
  } else {
    const current = join(runtime, "current");
    const currentPath = isSymlink(current) ? realpathSync(current) : undefined;
    const previousPath = join(runtime, "previous");
    const previous = existsSync(previousPath) ? readFileSync(previousPath, "utf8").trim() : undefined;
    const releases = join(runtime, "releases");
    if (existsSync(releases)) for (const name of readdirSync(releases)) {
      const path = join(releases, name);
      if (path !== currentPath && name !== previous) rmSync(path, { recursive: true });
    }
  }
  return undefined;
}

function option(args: string[], name: string): string {
  const index = args.indexOf(name); const value = args[index + 1];
  if (index < 0 || value === undefined || value.startsWith("--")) throw new Error("the following arguments are required: " + name);
  return value;
}
export function manageCli(args: string[]): number {
  try {
    const action = args[0];
    if (action !== "activate" && action !== "verify" && action !== "rollback" && action !== "clean") throw new Error("action must be one of: activate, verify, rollback, clean");
    const output = manageProjectRuntime(action, option(args, "--runtime-root"), option(args, "--project-root"));
    if (output !== undefined) console.log(output);
    return 0;
  } catch (error) { console.error(error instanceof Error ? error.message : String(error)); return 1; }
}
if (process.argv[1] !== undefined && basename(process.argv[1]).replace(/\.(?:js|ts)$/, "") === "manage-project-runtime") process.exitCode = manageCli(process.argv.slice(2));
